File: GUIController.m
#import "GUIController.h"
#import "AddGramsValueTransformer.h"
#import "AddMGValueTransformer.h"
#import "BeFitGraphOpenGLView.h"
#import "mySheetController.h"
#import "iTableColumnHeaderCell.h"
#import "INAppStoreWindow.h"
#import "INWindowButton.h"
#import <WebKit/WebKit.h>

@implementation GUIController
@synthesize window;
@synthesize webView;




- (void)setupCloseButton {
    INWindowButton *closeButton = [INWindowButton windowButtonWithSize:NSMakeSize(14, 16) groupIdentifier:nil];
    closeButton.activeImage = [NSImage imageNamed:@"close-active-color.tiff"];
    closeButton.activeNotKeyWindowImage = [NSImage imageNamed:@"close-activenokey-color.tiff"];
    closeButton.inactiveImage = [NSImage imageNamed:@"close-inactive-disabled-color.tiff"];
    closeButton.pressedImage = [NSImage imageNamed:@"close-pd-color.tiff"];
    closeButton.rolloverImage = [NSImage imageNamed:@"close-rollover-color.tiff"];
    self.window.closeButton = closeButton;
}

- (void)setupMinimizeButton {
    INWindowButton *button = [INWindowButton windowButtonWithSize:NSMakeSize(14, 16) groupIdentifier:nil];
    button.activeImage = [NSImage imageNamed:@"minimize-active-color.tiff"];
    button.activeNotKeyWindowImage = [NSImage imageNamed:@"minimize-activenokey-color.tiff"];
    button.inactiveImage = [NSImage imageNamed:@"minimize-inactive-disabled-color.tiff"];
    button.pressedImage = [NSImage imageNamed:@"minimize-pd-color.tiff"];
    button.rolloverImage = [NSImage imageNamed:@"minimize-rollover-color.tiff"];
    self.window.minimizeButton = button;
}

- (void)setupZoomButton {
    INWindowButton *button = [INWindowButton windowButtonWithSize:NSMakeSize(14, 16) groupIdentifier:nil];
    button.activeImage = [NSImage imageNamed:@"zoom-active-color.tiff"];
    button.activeNotKeyWindowImage = [NSImage imageNamed:@"zoom-activenokey-color.tiff"];
    button.inactiveImage = [NSImage imageNamed:@"zoom-inactive-disabled-color.tiff"];
    button.pressedImage = [NSImage imageNamed:@"zoom-pd-color.tiff"];
    button.rolloverImage = [NSImage imageNamed:@"zoom-rollover-color.tiff"];
    self.window.zoomButton = button;
}






- (IBAction)ShowHideNutritionPanel:(id)sender
{
	
	//Change whatever the current setting is for the nutrition panel's hide/show value
	[ NutritionPanelWinNSView setHidden: ![ NutritionPanelWinNSView isHidden ] ];
	
	//Show/Hide the nutrition panel
	NSRect CurrentFrame = [ FoodNSTableView frame ];
	
	if ( [ NutritionPanelWinNSView isHidden ] )
	{
		NSRect FrameToUse = CurrentFrame;
		
		FrameToUse.size.width = FrameToUse.size.width + [ NutritionPanelWinNSView frame ].size.width;
		[ FoodNameColumn setWidth: [ FoodNameColumn width ] + [ NutritionPanelWinNSView frame ].size.width ];
		
		[ FoodNSTableView setFrame: FrameToUse ];
		
		NSRect GraphFrame = [ GraphPanelWinNSView frame ];
		GraphFrame.size.width = GraphFrame.size.width + [ NutritionPanelWinNSView frame ].size.width;
		[ GraphPanelWinNSView setFrame: GraphFrame ];
		
		NSRect OpenGLFrame = [ webView frame ];
		OpenGLFrame.size.width = GraphFrame.size.width;
		[ GraphPanelWinNSView setFrame: OpenGLFrame ];
        
	}
	else
	{
		NSRect FrameToUse = CurrentFrame;
		
		FrameToUse.size.width = FrameToUse.size.width - [ NutritionPanelWinNSView frame ].size.width;
		[ FoodNameColumn setWidth: [ FoodNameColumn width ] - [ NutritionPanelWinNSView frame ].size.width ];
		
		[ FoodNSTableView setFrame: FrameToUse ];
		
		NSRect GraphFrame = [ GraphPanelWinNSView frame ];
		GraphFrame.size.width = GraphFrame.size.width - [ NutritionPanelWinNSView frame ].size.width;
		[ GraphPanelWinNSView setFrame: GraphFrame ];
        
		NSRect OpenGLFrame = [ webView frame ];
		OpenGLFrame.size.width = GraphFrame.size.width;
		[ GraphPanelWinNSView setFrame: OpenGLFrame ];
        
	}
}


-(BOOL)changeableColumnIsEditable
{
	return( NO );
}



- (void)tableViewSelectionDidChange:(NSNotification *)aNotification
{
	
	//Sent when the food list changes we need to redraw the individual food table here to avoid drawing artifacts
	//Really this shouldn't be needed but seems Mac OS X might have a bug here at least on Tiger.
	[ FoodNSTableView setNeedsDisplay: YES ];
	
	if( [ FoodListsNSTableView selectedRow ] == 0 )
	{
		[ FoodQuantityTableColumn setHidden: YES ];
        [ FoodRatingTableColumn setHidden: YES ];
        [ FoodDateTableColumn setHidden: YES ];
	} else {
		[ FoodQuantityTableColumn setHidden: NO ];
        [ FoodRatingTableColumn setHidden: NO ];
        [ FoodDateTableColumn setHidden: YES ];
	}
    
}

-(void)doNothing
{
	
	//What did you think this did?
}



-(void)ShowGraphPanel
{
	
	//Show the graph panel
	if ( [ GraphPanelWinNSView isHidden ] )
	{ [ self ShowHideGraphPanel: self ]; }
}



-(void)HideGraphPanel
{
	
	//Hide the graph panel
	if ( ![ GraphPanelWinNSView isHidden ] )
	{ [ self ShowHideGraphPanel: self ]; }
}


-(void)awakeFromNib
{
    
    //Set Done Button Color
    
    NSColor *color = [NSColor whiteColor];
    
    NSMutableAttributedString *colorTitle =
    [[NSMutableAttributedString alloc] initWithAttributedString:[Finishedbutton attributedTitle]];
    
    NSRange titleRange = NSMakeRange(0, [colorTitle length]);
    
    [colorTitle addAttribute:NSForegroundColorAttributeName
                       value:color
                       range:titleRange];
    
    [Finishedbutton setAttributedTitle:colorTitle];
    
    
    NSScrollView *sv = [FoodNSTableView enclosingScrollView];
    [sv setScrollerStyle: NSScrollerStyleOverlay];
    
    NSString *path1 = [[NSBundle mainBundle] pathForResource:@"index" ofType:@"html"];
    NSString *html = [NSString stringWithContentsOfFile:path1 encoding:NSUTF8StringEncoding error:nil];
    [[[webView mainFrame] frameView] setAllowsScrolling:NO];
    [[webView mainFrame] loadHTMLString:html baseURL:[[NSBundle mainBundle] resourceURL]];
    [webView setDrawsBackground:NO];
    
    NSString *path2 = [[NSBundle mainBundle] pathForResource:@"foodgraph" ofType:@"html"];
    NSString *html2 = [NSString stringWithContentsOfFile:path2 encoding:NSUTF8StringEncoding error:nil];
    [[[lowerGraphView mainFrame] frameView] setAllowsScrolling:NO];
    [[lowerGraphView mainFrame] loadHTMLString:html2 baseURL:[[NSBundle mainBundle] resourceURL]];
    [lowerGraphView setDrawsBackground:NO];
    
    
    /* set preference defaults */
    [[NSUserDefaults standardUserDefaults] registerDefaults:
     [NSDictionary dictionaryWithObject: [NSNumber numberWithBool: YES]
                                 forKey: @"NSDisabledCharacterPaletteMenuItem"]];
    
    self.window.trafficLightButtonsLeftMargin = 10.0;
    self.window.centerTrafficLightButtons = NO;
    self.window.centerFullScreenButton = NO;
    self.window.titleBarHeight = 80.0;
    self.window.trafficLightButtonsTopMargin = 10;
    self.window.fullScreenButtonRightMargin = 5;
    self.window.fullScreenButtonTopMargin = 5;
    self.window.showsTitle = YES;
    self.titleView.frame = self.window.titleBarView.bounds;
    self.titleView.autoresizingMask = NSViewWidthSizable | NSViewHeightSizable;
    [self.window.titleBarView addSubview:self.titleView];
    
    [self setupCloseButton];
    [self setupMinimizeButton];
    [self setupZoomButton];
    
    
    
    
    
    [tableView reloadData];
    
    NSRect frame = tableView.headerView.frame;
    frame.size.height = 17;
    tableView.headerView.frame = frame;
	
    NSArray *columns = [tableView tableColumns];
    NSEnumerator *cols = [columns objectEnumerator];
    NSTableColumn *col = nil;
    
    iTableColumnHeaderCell *iHeaderCell;
    
    while (col = [cols nextObject]) {
        iHeaderCell = [[iTableColumnHeaderCell alloc]
                       initTextCell:[[col headerCell] stringValue]];
        [col setHeaderCell:iHeaderCell];
        [iHeaderCell release];
    }
	
	
	[mainWindow setDelegate:self];
    
	
	//Setting this so we can disable items in the menu
	[ FileMenu setAutoenablesItems: NO ];
	
	
	//Add the nutrition panel to the window
    [myScrollView setDocumentView:finalView];
    
    NSPoint pt = NSMakePoint(0.0, [[myScrollView documentView]
                                   bounds].size.height);
    [[myScrollView documentView] scrollPoint:pt];
    
	
    
	[ NutritionPanelWinNSView setNeedsDisplay: TRUE ];
	
	
	[ GraphPanelWinNSView setNeedsDisplay: TRUE ];
	
	//Register the value transformers
	AddGramsValueTransformer* GramsTransformer = [ [ [ AddGramsValueTransformer alloc ] init ] autorelease ];
	
	[ NSValueTransformer setValueTransformer: GramsTransformer forName: @"addGrams" ];
	
	AddMGValueTransformer* MGTransformer = [ [ [ AddMGValueTransformer alloc ] init ] autorelease ];
	
	[ NSValueTransformer setValueTransformer: MGTransformer forName: @"addMG" ];
	
	[ self HideGraphPanel ];
    
    
    
    
    
    
    //Add the nutrition panel to the window
	
	
	
    
	[myBackView setDocumentView:finalBackView];
    
    
    [myBackView setBackgroundColor:[NSColor colorWithPatternImage:[NSImage imageNamed:@"black_linen_v2.png"]]];
    
    
}

- (void)windowWillClose:(NSNotification *)aNotification {
	[NSApp terminate:self];
}




-(IBAction)loadSupport:(id)sender{
	
	NSURL *url=[NSURL URLWithString:@"http://www.jonbrown.org/support/#bfit"];
	[[NSWorkspace sharedWorkspace] openURL:url];
	
}





-(IBAction)loadEmaildev:(id)sender{
	NSURL *url=[NSURL URLWithString:@"mailto:jon@jonbrown.org?subject=BeFit%20Feedback&body=Feedback"];
	[[NSWorkspace sharedWorkspace] openURL:url];
}



-(IBAction)loadWebsite:(id)sender{
	NSURL *url=[NSURL URLWithString:@"http://osx.iusethis.com/app/befit"];
	[[NSWorkspace sharedWorkspace] openURL:url];
}


-(IBAction)loadFacebook:(id)sender{
	NSURL *url=[NSURL URLWithString:@"http://www.facebook.com/sharer.php?s=100&p[url]=http://www.jonbrown.org/befit/&p[title]=BeFit&p[summary]=BeFit+is+the+ultimate+calorie+tracking+tool.+Based+on+the+nutrition+data+for+the+foods+that+you+eat.+Enter+the+coupon+code+%23BF2014+and+get+BeFit+at+30%25+off+for+a+limited+time%21"];
	[[NSWorkspace sharedWorkspace] openURL:url];
}

-(IBAction)loadTwitter:(id)sender{
	NSURL *url=[NSURL URLWithString:@"http://twitter.com/share?url=http%3A%2F%2Fwww.jonbrown.org%2Fbefit%2F&via=JonBrownDesigns&text=BeFit+is+the+ultimate+calorie+tracking+tool.+Based+on+the+nutrition+data+for+the+foods+that+you+eat.+Enter+the+coupon+code+%23BF2014+and+get+BeFit+at+30%25+off+for+a+limited+time%21"];
	[[NSWorkspace sharedWorkspace] openURL:url];
}

- (IBAction)ShowHideGraphPanel:(id)sender
{
	
	//Change whatever the current setting is for the graph panel's hide/show value
	[ GraphPanelWinNSView setHidden: ![ GraphPanelWinNSView isHidden ] ];
	
	//Show/Hide the graph panel
	NSRect CurrentFrame = [ FoodNSTableView frame ];
	
	//We don't adjust the full height since one pixel is hidden of the graph panel
	float HeightToAdjust = [ GraphPanelWinNSView frame ].size.height - 1;
	
	if ( [ GraphPanelWinNSView isHidden ] )
	{
		NSRect FrameToUse = CurrentFrame;
		
		FrameToUse.size.height = FrameToUse.size.height + HeightToAdjust;
		FrameToUse.origin.y = FrameToUse.origin.y - HeightToAdjust;
		
		[ FoodNSTableView setFrame: FrameToUse ];
	}
	else
	{
		NSRect FrameToUse = CurrentFrame;
		
		FrameToUse.size.height = FrameToUse.size.height - HeightToAdjust;
		FrameToUse.origin.y = FrameToUse.origin.y + HeightToAdjust;
		
		[ FoodNSTableView setFrame: FrameToUse ];
	}
	
	[ FoodNSTableView setNeedsDisplay: TRUE ];
	[ GraphPanelWinNSView setNeedsDisplay: TRUE ];
}


- (IBAction)print:(id)sender {
    
    NSPrintOperation* printOperation = [NSPrintOperation printOperationWithView:PrintView];
    [printOperation setCanSpawnSeparateThread:YES];
    [printOperation runOperationModalForWindow:[self window] delegate:MainWindow didRunSelector:nil contextInfo:nil];
    
}

- (IBAction)pagesetup:(id)sender {
    
    
    NSPrintInfo *printInfo = [NSPrintInfo sharedPrintInfo];
    NSPageLayout *pageLayout = [NSPageLayout pageLayout];
    
    
    [pageLayout beginSheetWithPrintInfo:printInfo modalForWindow:[self window] delegate:MainWindow didEndSelector:@selector(pageLayoutDidEnd:returnCode:contextInfo:) contextInfo:nil];
    
    
    //runs the model for the page layout UI.  It saves the global copy of printInfo in printOperation, which can be used to make decisions
    
    
}

- (void)pageLayoutDidEnd:(NSPageLayout *)pageLayout returnCode:(int)returnCode contextInfo:(void *)contextInfo
{
    if (returnCode == NSOKButton)
    {
    }
}


- (IBAction)ResizeCols:(id)sender {
    
    [tableView setColumnAutoresizingStyle:NSTableViewUniformColumnAutoresizingStyle];
    
    [FoodNameColumn setResizingMask:NSTableColumnUserResizingMask];
    [CalNameCalumn setResizingMask:NSTableColumnUserResizingMask];
    [MixNameColumn setResizingMask:NSTableColumnUserResizingMask];
    [FoodQuantityTableColumn setResizingMask:NSTableColumnUserResizingMask];
    [FoodRatingTableColumn setResizingMask:NSTableColumnUserResizingMask];
    [FoodDateTableColumn setResizingMask:NSTableColumnUserResizingMask];
    
    //[FoodNameColumn sizeToFit];
    [CalNameCalumn setWidth:70];
    [MixNameColumn setWidth:70];
    [FoodQuantityTableColumn setWidth:70];
    [FoodRatingTableColumn setWidth:100];
    [FoodDateTableColumn setWidth:100];
    
    [FoodNameColumn setWidth:400];
}



@end

File: GUI Objects/GUIController.swift
//
//  GUIController.swift
//  BeFit
//
//  Created by Andrej Slegl on 11/5/19.
//

import Foundation
import Cocoa
import WebKit
import AppKit

class GUIController: NSObject, NSTableViewDelegate, NSWindowDelegate {
    @IBOutlet var FileMenu: NSMenu!
    @IBOutlet var FoodListsNSTableView: NSTableView!
    @IBOutlet var FoodNameColumn: NSTableColumn!
    @IBOutlet var MixNameColumn: NSTableColumn!
    @IBOutlet var CalNameCalumn: NSTableColumn!
    @IBOutlet var FoodNSTableView: NSTableView!
    @IBOutlet var FoodQuantityTableColumn: NSTableColumn!
    @IBOutlet var FoodRatingTableColumn: NSTableColumn!
    @IBOutlet var GraphPanelWinNSView: NSView!
    @IBOutlet var mainWindow: AnyObject!
    @IBOutlet var NutritionPanelDetailWinNSView: NSView!
    @IBOutlet var tableView2: NSTableView!
    @IBOutlet var NutritionPanelWinBackgroundNSView: NSView!
    @IBOutlet var PrintView: NSView!
    @IBOutlet var pieIndex1GraphView: WebView!
    @IBOutlet var foodGraphWebView: WebView!
    @IBOutlet var pieIndex2GraphView: WebView!
    @IBOutlet var pieIndex3GraphView: WebView!
    @IBOutlet var guageWebView: WebView!
    @IBOutlet var horizontalwebView: WebView!
    @IBOutlet var window: NSWindow!
    @IBOutlet var healthPopover: NSPopover!
    @IBOutlet var healthPopoverQtyOnly: NSPopover!
    @IBOutlet var topToolbar: NSToolbar!
    @IBOutlet var frontViewScrollView: NSScrollView!
    @IBOutlet var frontView: NSView!
    @IBOutlet var backView: NSView!
    @IBOutlet var showHideButton: NSButton!
    
    private func loadHTMLString(fromResource resource: String, to view: WebView?) {
        guard let path = Bundle.main.path(forResource: resource, ofType: "html") else {
            assertionFailure()
            return
        }
        
        guard let html = try? String(contentsOfFile: path, encoding: .utf8) else {
            assertionFailure()
            return
        }
        
        view?.mainFrame.frameView.allowsScrolling = false
        view?.mainFrame.loadHTMLString(html, baseURL: Bundle.main.resourceURL)
        view?.drawsBackground = false
    }
    
    override func awakeFromNib() {
        super.awakeFromNib()
        
        let sv = FoodNSTableView.enclosingScrollView
        sv?.scrollerStyle = .overlay
        var frame = tableView2.headerView?.frame ?? .zero
        frame.size.height = 26
        tableView2.headerView?.frame = frame

        pieIndex1GraphView.setValue(false, forKey: "drawsBackground")
        pieIndex2GraphView.setValue(false, forKey: "drawsBackground")
        pieIndex3GraphView.setValue(false, forKey: "drawsBackground")
        guageWebView.setValue(false, forKey: "drawsBackground")
        foodGraphWebView.setValue(false, forKey: "drawsBackground")
        horizontalwebView.setValue(false, forKey: "drawsBackground")
        
        // Is Dark Mode Enabled?
        enum InterfaceStyle : String {
           case Dark, Light

           init() {
              let type = UserDefaults.standard.string(forKey: "AppleInterfaceStyle") ?? "Light"
              self = InterfaceStyle(rawValue: type)!
            }
        }
        
        let currentStyle = InterfaceStyle()
        
        // Set graphing system
        if currentStyle.rawValue == "Dark" {
            loadHTMLString(fromResource: "maingraph_dark", to: horizontalwebView)
            loadHTMLString(fromResource: "pieindex_dark", to: pieIndex1GraphView)
            loadHTMLString(fromResource: "pieindex2_dark", to: pieIndex2GraphView)
            loadHTMLString(fromResource: "pieindex3_dark", to: pieIndex3GraphView)
            loadHTMLString(fromResource: "foodgraph_dark", to: foodGraphWebView)
            loadHTMLString(fromResource: "gauge_dark", to: guageWebView)
        } else {
            
            loadHTMLString(fromResource: "maingraph", to: horizontalwebView)
            loadHTMLString(fromResource: "pieindex", to: pieIndex1GraphView)
            loadHTMLString(fromResource: "pieindex2", to: pieIndex2GraphView)
            loadHTMLString(fromResource: "pieindex3", to: pieIndex3GraphView)
            loadHTMLString(fromResource: "foodgraph", to: foodGraphWebView)
            loadHTMLString(fromResource: "gauge", to: guageWebView)
        }

        // Set preference defaults
        UserDefaults.standard.register(defaults: ["NSDisabledCharacterPaletteMenuItem": NSNumber(value: true)])
                
        tableView2.reloadData()
        
        // Set height of the TableHeader
        var frame_head = tableView2.headerView?.frame ?? .zero
        frame_head.size.height = 20
        tableView2.headerView?.frame = frame_head
        
        //Setting this so we can disable items in the menu
        FileMenu.autoenablesItems = false
        
        //Set Flip Value
        UserDefaults.standard.set(false, forKey: "flipPref")
        
        //Add the nutrition panel to the window
        
        frontViewScrollView.documentView?.setFrameSize(NSMakeSize(205, 1511))
        NutritionPanelWinBackgroundNSView.addSubview(frontViewScrollView)
        frontViewScrollView.documentView?.addSubview(frontView)
        
        if let documentView = frontViewScrollView.documentView {
                documentView.scroll(NSPoint(x: 0, y: documentView.bounds.size.height))
        }
        
        
        NutritionPanelWinBackgroundNSView.needsDisplay = true
        GraphPanelWinNSView.needsDisplay = true
        
        //Register the value transformers
        ValueTransformer.setValueTransformer(AddGramsValueTransformer(), forName: AddGramsValueTransformer.name)
        ValueTransformer.setValueTransformer(AddMGValueTransformer(), forName: AddMGValueTransformer.name)
        
        HideGraphPanel()
        
    }
    
    // Make some columns non-editable
    @objc func changeableColumnIsEditable() -> Bool {
        return false
    }
    
    // Important sets the columns as hidden or visible based on which one is selected. GuiController must be the delegate of the SourceList for this to work.
    func tableViewSelectionDidChange(_ notification: Notification) {
        //Sent when the food list changes we need to redraw the individual food table here to avoid drawing artifacts
        //Really this shouldn't be needed but seems Mac OS X might have a bug here at least on Tiger.
        FoodNSTableView.needsDisplay = true
        
        if FoodListsNSTableView.selectedRow == 0 || FoodListsNSTableView.selectedRow == 1 {
            FoodQuantityTableColumn.isHidden = true
            FoodRatingTableColumn.isHidden = false
        } else {
            FoodQuantityTableColumn.isHidden = false
            FoodRatingTableColumn.isHidden = false
        }
    }
    
    // Make some libraries non-editable
    func tableView(_ tableView: NSTableView, viewFor tableColumn: NSTableColumn?, row: Int) -> NSView? {
        guard let column = tableColumn else {
            return nil
        }
        
        guard let cellView = tableView.makeView(withIdentifier: column.identifier, owner: self) as? NSTableCellView else {
            return nil
        }
        
        let textField = cellView.textField
        textField?.isEditable = false
        return cellView
    }
    
    // Helper Functions for the Actions below
    func ShowGraphPanel() {
        if GraphPanelWinNSView.isHidden {
            ShowHideGraphPanel(self)
        }
    }
    
    func HideGraphPanel() {
        if !GraphPanelWinNSView.isHidden {
            ShowHideGraphPanel(self)
        }
    }
    
    // Quit the App if it is closed by pressing the red close button.
    func windowWillClose(_ notification: Notification) {
        NSApp.terminate(self)
    }
    
    @IBAction func ShowHideGraphPanel(_ sender: AnyObject) {
        //Change whatever the current setting is for the graph panel's hide/show value
        GraphPanelWinNSView.isHidden = !GraphPanelWinNSView.isHidden
        
        //Show/Hide the graph panel
        let currentFrame = FoodNSTableView.frame
        
        //We don't adjust the full height since one pixel is hidden of the graph panel
        let heightToAdjust = GraphPanelWinNSView.frame.size.height - 1
        
        if GraphPanelWinNSView.isHidden {
            var frameToUse = currentFrame
            frameToUse.size.height = frameToUse.size.height + heightToAdjust
            frameToUse.origin.y = frameToUse.origin.y - heightToAdjust

            FoodNSTableView.frame = frameToUse
        } else {
            var frameToUse = currentFrame
            frameToUse.size.height = frameToUse.size.height - heightToAdjust
            frameToUse.origin.y = frameToUse.origin.y + heightToAdjust

            FoodNSTableView.frame = frameToUse
        }

        FoodNSTableView.needsDisplay = true
        GraphPanelWinNSView.needsDisplay = true
    }
    
    @IBAction func print(_ sender: AnyObject) {
        let printInfo = NSPrintInfo.shared
        printInfo.verticalPagination = .automatic
        let horizontalMargin: CGFloat = -100
        let verticalMargin: CGFloat = -100
        
        printInfo.leftMargin = horizontalMargin
        printInfo.rightMargin = horizontalMargin
        printInfo.isHorizontallyCentered = true
        printInfo.topMargin = -600
        printInfo.bottomMargin = verticalMargin
        
        let printOperation = NSPrintOperation(view: PrintView)
        printOperation.canSpawnSeparateThread = true
        printOperation.runModal(for: window, delegate: nil, didRun: nil, contextInfo: nil)
    }
    
    @IBAction func pagesetup(_ sender: AnyObject) {
        let printInfo = NSPrintInfo.shared
        let pageLayout = NSPageLayout()
        pageLayout.beginSheet(with: printInfo, modalFor: window, delegate: nil, didEnd: nil, contextInfo: nil)
        //runs the model for the page layout UI.  It saves the global copy of printInfo in printOperation, which can be used to make decisions
    }
    
    @IBAction func ResizeCols(_ sender: AnyObject) {
        tableView2.columnAutoresizingStyle = .uniformColumnAutoresizingStyle
        FoodNameColumn.resizingMask = .userResizingMask
        CalNameCalumn.resizingMask = .userResizingMask
        MixNameColumn.resizingMask = .userResizingMask
        FoodQuantityTableColumn.resizingMask = .userResizingMask
        FoodRatingTableColumn.resizingMask = .userResizingMask
        
        CalNameCalumn.width = 70
        MixNameColumn.width = 70
        FoodQuantityTableColumn.width = 70
        FoodRatingTableColumn.width = 100
        FoodNameColumn.width = 300
    }
    
    @IBAction func ShowHideNutritionPanel(_ sender: AnyObject) {
        //Change whatever the current setting is for the nutrition panel's hide/show value
        NutritionPanelWinBackgroundNSView.isHidden = !NutritionPanelWinBackgroundNSView.isHidden
        //backView.isHidden = !backView.isHidden
        NutritionPanelDetailWinNSView.isHidden = !NutritionPanelDetailWinNSView.isHidden
        
        let currentFrame = FoodNSTableView.frame
        
        
        if NutritionPanelWinBackgroundNSView.isHidden {
            
            showHideButton.title = "Show"
            var frameToUse = currentFrame
            frameToUse.size.width = frameToUse.size.width + NutritionPanelWinBackgroundNSView.frame.size.width
            FoodNameColumn.width = FoodNameColumn.width + NutritionPanelWinBackgroundNSView.frame.size.width
                        
            frameToUse.size.width = frameToUse.size.width + NutritionPanelDetailWinNSView.frame.size.width
            FoodNameColumn.width = FoodNameColumn.width + NutritionPanelDetailWinNSView.frame.size.width
            
            FoodNSTableView.frame = frameToUse
            
            var graphFrame = GraphPanelWinNSView.frame
            graphFrame.size.width = graphFrame.size.width + NutritionPanelWinBackgroundNSView.frame.size.width
            GraphPanelWinNSView.frame = graphFrame

            var openGLFrame = horizontalwebView.frame
            openGLFrame.size.width = graphFrame.size.width
            GraphPanelWinNSView.frame = openGLFrame
            
            
        } else {
            
            showHideButton.title = "Hide"
            var frameToUse = currentFrame
            frameToUse.size.width = frameToUse.size.width - NutritionPanelWinBackgroundNSView.frame.size.width
            FoodNameColumn.width = FoodNameColumn.width - NutritionPanelWinBackgroundNSView.frame.size.width

            frameToUse.size.width = frameToUse.size.width - NutritionPanelWinBackgroundNSView.frame.size.width
            FoodNameColumn.width = FoodNameColumn.width - NutritionPanelWinBackgroundNSView.frame.size.width
            
            FoodNSTableView.frame = frameToUse
            
            var graphFrame = GraphPanelWinNSView.frame
            graphFrame.size.width = graphFrame.size.width - NutritionPanelWinBackgroundNSView.frame.size.width
            GraphPanelWinNSView.frame = graphFrame
            
            var openGLFrame = horizontalwebView.frame
            openGLFrame.size.width = graphFrame.size.width
            GraphPanelWinNSView.frame = openGLFrame
        }
    }
    
}
